Sprint 14

bakgeldia opened the pull request:
Задание 14-спринта:

1. Создан новый проект Tracker, в котором
- iOS 13.4 указана как минимально поддерживаемая версия; 
- для проекта подключён Git.
2. Вёрстка экранов, перечисленных ниже, полностью соответствует макету
3. Свёрстан Launch Screen приложения
4. На главном экране — две вкладки: «Трекеры» и «Статистика». Вторая вкладка пока пустая.
5. Свёрстан главный экран «Трекеры»
6. Реализованы экраны создания трекера
7. Реализован экран задания расписания для трекера

bakgeldia committed:
- добавлены структуры Tracker, TrackerCategory и TrackerRecord
- кнока + и "datePicker" были добавлены как NavBarItem

bakgeldia committed:
Добавлены AddTrackerViewController, NewHabitViewController
- При нажатии кнопки "+" на экране Трекеры, появляется шторка экрана "Создание трекера" с кнопками "Привычка" и "Нерегулярное событие"
- При нажатии на кнопку "Привычка", появляется шторка экрана "Новая привычка" с текстовым поле и таблицей с 2 ячейками "Категория" и "Расписание"

bakgeldia committed:
На экране "Новая привычка" добавлены кнопки "Отменить" и "Создать".
При нажатии на кнопку "Отменить" экран закрывается и возвращается обратно на экран "Создание трекера"

bakgeldia committed:
Добавлен экран Расписание.
- При нажатии на "Расписание" на экране "Новая привычка", открывается шторка экрана Расписание
- На экране Расписание добавлена таблица с 7 ячейками с названием дни недели в каждом и переключателем(switcher) справа.
- Внизу экрана добавлена кнопка "Готово".  Пока при нажатии на нее, закрывается экран Расписание.

bakgeldia committed:
- SearchBar и заголовок экрана Трекеры помещены в NavBar
- Добавлен CollectionView

bakgeldia committed:
Добавлены:
- Фильтрация по поисковику и даты
- Сверстан экран создания Нерегулярного события
- Трекеры распределены по категориям на экране

Andreynnt commented on Tracker/Assets.xcassets/Done.imageset/Done.png:
Картинки 1x лучше не добавлять. Они нужны были для iPhone без retina экранов, а последний такой - iPhone5. Он не поддерживает минимальную версию учебного приложения (iOS 13), поэтому можно смело удалять, чтобы не занимать лишнюю память

Andreynnt commented on Tracker/Trackers/CategoryHeaderReusableView.swift:
```diff
@@ -0,0 +1,34 @@
+
+import UIKit
+
+final class CategoryHeaderReusableView: UICollectionReusableView {
```
Молодец, что не забываешь про final!

Andreynnt commented on Tracker/Trackers/TrackerCollectionViewCell.swift:
```diff
@@ -0,0 +1,132 @@
+
+final class TrackerCollectionViewCell: UICollectionViewCell {
+    let emojiAndNameView = UIView()
+    let emoji = UILabel()
```
Можно добавить private к свойствам

Andreynnt commented on Tracker/Trackers/TrackerCreation/HabitCreation/CategoryViewController.swift:
```diff
@@ -0,0 +1,113 @@
+extension CategoryViewController: UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        // Возвращаем количество категорий
+        return categories.count
```
В однострочных функциях/вычисляемых свойствах return можно не писать)

Andreynnt commented on Tracker/Trackers/TrackerCreation/NonRegularEvent/NonReguarEventViewController.swift:
```diff
@@ -0,0 +1,173 @@
+        categoryVC.modalPresentationStyle = .popover
+        categoryVC.categories = self.categories
+        
+        self.present(categoryVC, animated: true, completion: nil)
```
self лучше не писать, если нет замыкания

bakgeldia committed:
Исправления после код ревью №1:
- Удалены картинки 1х
- Кнопка + деактивируется на будущие даты
- При отметке трекера он добавляется/удаляется в/из массива completedTrackers

bakgeldia committed:
Исправления после код ревью №2:
- Добавлена возможность выбора категории для привычки/нерегулярного события
- Добавлена возможность выбора расписания для привычки
- Добавлена возможность создания привычки/нерегулярного события
- Исправлены скругления у таблицы на экранах Категория и Расписания
- При создании трекера, он отображается в главном экране Трекеров, и показывается при фильтрации по поиску и дате.

Andreynnt commented:
<img width="356" alt="image" src="https://github.com/user-attachments/assets/c54ecbbd-d1f9-45dc-9ae9-a5cee9d629e8">
<img width="327" alt="image" src="https://github.com/user-attachments/assets/29e64e5a-223d-45e1-88d5-ccd71d498971">

bakgeldia committed:
Иcправления после код ревью №3:
- Добавлена круглая обводка для эмодзи трекера
- Nav bar и tab bar теперь непрозрачные, чтобы при скролле списка трекеров, они не показывались внутри нав бара и таб бара
- Удалены нижние  сепараторы последних ячеек в таблицах (на странице создания привычки/события, категория
- Размеры шрифтов установлены по дизайну

bakgeldia committed:
Исправления после код ревью №4:
- Исправлен отступ названия секции слева
- При создании трекера показывался весь список трекеров. Исправлено.
- После создания трекера экраны "выбора привычки или события" и "создание трекера" закрываются одновременно.

bakgeldia committed:
Удалена пустая папка

bakgeldia committed:
Исправлен UISearchBar:
- Текст и цвет кнопки "Отменить"
- Элементы навбара остаются на месте при поиске

bakgeldia committed:
Сохраняются выбранные дни при закрытии и повторном открытии "Расписание"
